File: ios/AIUsageMobile/PairingCodeViewModel.swift
import Foundation

enum PairingCodeState: Equatable {
    case idle
    case loading
    case ready(code: String, expiresAt: Date)
    case failed(message: String)

    var title: String {
        switch self {
        case .idle:
            return "No PC linked"
        case .loading:
            return "Generating code"
        case .ready:
            return "Your PC link code"
        case .failed:
            return "Could not generate code"
        }
    }
}

@MainActor
final class PairingCodeViewModel: ObservableObject {
    @Published private(set) var state: PairingCodeState = .idle

    private let apiClient: AIUsageAPIClient

    init(apiClient: AIUsageAPIClient) {
        self.apiClient = apiClient
    }

    func generate(idToken: String) async {
        state = .loading
        do {
            let response = try await apiClient.createPairingCode(idToken: idToken)
            state = .ready(code: response.displayCode, expiresAt: response.expiresAt)
        } catch {
            state = .failed(message: "Expires in code generation failed")
        }
    }

    func countdownText(now: Date = Date()) -> String? {
        guard case let .ready(_, expiresAt) = state else {
            return nil
        }
        let remaining = max(0, Int(expiresAt.timeIntervalSince(now)))
        return "Expires in \(remaining / 60):\(String(format: "%02d", remaining % 60))"
    }
}

File: android/app/src/main/java/com/aiusage/mobile/sync/PairingCodeViewModel.kt
package com.aiusage.mobile.sync

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface PairingCodeUiState {
    data object Idle : PairingCodeUiState
    data object Loading : PairingCodeUiState
    data class Ready(val code: String, val expiresAt: String, val helperText: String) : PairingCodeUiState
    data class Failed(val message: String) : PairingCodeUiState
}

class PairingCodeViewModel(
    private val apiClient: AIUsageApiClient
) : ViewModel() {
    private val mutableState = MutableStateFlow<PairingCodeUiState>(PairingCodeUiState.Idle)
    val state: StateFlow<PairingCodeUiState> = mutableState.asStateFlow()

    fun generate(idToken: String) {
        mutableState.value = PairingCodeUiState.Loading
        viewModelScope.launch {
            try {
                val response = apiClient.createPairingCode(idToken)
                mutableState.value = PairingCodeUiState.Ready(
                    code = response.displayCode,
                    expiresAt = response.expiresAt,
                    helperText = "Expires in 10:00"
                )
            } catch (error: Throwable) {
                mutableState.value = PairingCodeUiState.Failed("Could not generate PC link code")
            }
        }
    }
}

